### PR TITLE
Added bounded sharing using greedy coloring

### PR DESCRIPTION
**Pull request by calebmkim**
Implements bounded sharing using greedy coloring. 

Currently, the bounds should be passed as an extra command line argument, such as `-x cell-share:bounds=4,7,10`. If there is no such argument passed, default values of 4,6, and 18 are used (this is very arbitrary). Edit: I've changed it so that by default it always shares.

### CODE REVIEW

**Comment by rachitnigam** [comment]
This is very cool! I'd recommend updating the doc comment for the `cell-share` pass to talk about this option and how it works. The defaults are a little arbitrary; I wonder if always sharing would be a better default? CC @sampsyo?

**Comment by calebmkim** [comment]
Yeah I agree about the defaults being arbitrary. I was thinking this could be a good long term-solution: once we get the lenet stuff working, we can look at the results for some of the large .futil files, and see which sharing settings seem to be optimal and pick based on that. But we can also add another command line option to always share if that's what the user wants to do.

File: calyx/src/passes/cell_share.rs [hunk 1]
```diff
@@ -35,6 +55,10 @@ pub struct CellShare {
 
     /// Cell active in continuous assignments, or ref cells (we want to ignore both)
     cont_ref_cells: HashSet<ir::Id>,
+
+    /// The number of times a given class of cell can be shared. bounds should be
+    /// length 3 to hold the 3 classes: comb cells, registers, and everything else
+    pub bounds: Option<Vec<u64>>,
```
Comment (rachitnigam) [inline calyx/src/passes/cell_share.rs:61]: Seems like this doesn't need to be `pub`

**Review by rachitnigam — APPROVED** [review]
LGTM!

**Comment by sampsyo** [comment]
Good question about what to do for the defaults! It seems like there's no particularly great choice; I like @calebmkim's goal of setting the defaults based on some empirical measurements eventually. Always sharing by default is also sensible, since at least it's not surprising.

**Pull request merged**